File: hash/good_partitions/good_partitions.py

def count_good_partitions(nums: list[int]) -> int:  # LeetCode Q.2963.
    nums2ranges = dict()  # Key: num; value: 1st and last occurred indices.
    for idx, num in enumerate(nums):
        if num not in nums2ranges.keys():
            nums2ranges.update({num: [idx, idx]})
        nums2ranges[num][1] = idx  # Update last occurred idx.

    total_segments = 0
    last_segment_first_idx, last_segment_last_idx = -1, -1
    for first_idx, last_idx in nums2ranges.values():
        if first_idx <= last_segment_last_idx:  # Merge current and last segments.
            last_segment_last_idx = max(last_segment_last_idx, last_idx)
            continue
        last_segment_first_idx, last_segment_last_idx = first_idx, last_idx
        total_segments += 1  # Isolate current segment.

    return (2 ** (total_segments - 1)) % (10 ** 9 + 7)  # Required to control size.

File: hash/good_partitions/good_partitions.cpp
#include <vector>
#include <unordered_map>
using namespace std;

int countGoodPartitions(vector<int> &nums)
{ // LeetCode Q.2963.
    // Each num's 1st and last occurred indices.
    unordered_map<int, pair<int, int>> nums2BoundaryIndices;

    for (int idx = 0; idx < nums.size(); idx++)
    {
        if (nums2BoundaryIndices.find(nums[idx]) == nums2BoundaryIndices.end())
            nums2BoundaryIndices[nums[idx]] = {idx, idx};

        nums2BoundaryIndices[nums[idx]].second = idx;
    }

    int segments = 0;
    int lastSegmentLeftIdx = -1, lastSegmentRightIdx = -1;

    for (int idx = 0; idx < nums.size(); idx++)
    {
        auto [leftIdx, rightIdx] = nums2BoundaryIndices[nums[idx]];
        if (lastSegmentRightIdx < leftIdx)
        { // A new segment is formed.
            segments += 1;
            lastSegmentLeftIdx = leftIdx, lastSegmentRightIdx = rightIdx;
            continue;
        }

        if (lastSegmentRightIdx < rightIdx) // Last segment expands.
            lastSegmentRightIdx = rightIdx;
    }

    long long partitions = 1, modulo = pow(10, 9) + 7;
    while (segments > 1)
    { // Partitions = 2 ** (segments - 1).
        partitions *= 2;
        partitions %= modulo; // Required to control size.
        segments--;
    }

    return partitions;
}
